### PR TITLE
Draft compact index documentation

### PR DESCRIPTION
**Pull request by martinemde**
Please offer fixes or clarification if I got any of this wrong.

### CODE REVIEW

File: rubygems-org-compact-index-api.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    [...]
+    active_model_serializers -0.9.10 7ad37af4aec8cc089e409e1fdec86f3d
+    active_model_serializers 0.9.11 a6d40e97b289ee6c806e5e9f7031623b
+    openapi_first 1.4.1 40fbfdebcbfee3863df697e1d641f637
+
+#### `versions` File Format
+
+The format of the `versions` file uses one line per gem at computation time, with additional lines appended to the end that may include new or yanked versions of a gem already present earlier in the file.
+
+The first line is a created_at timestamp in IOS8601 format.
```
Comment (segiddins) [inline rubygems-org-compact-index-api.md:84]: I think for all compact index files, everything up to the `---` is opaque metadata. Furthermore, the opaque metadata and `---` separator line are optional.

File: rubygems-org-compact-index-api.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+The parts of this line are as follows:
+
+1. **`GEM`** - The name of the gem.
+2. **`(SPACE)`** - The space character.
+3. **`[(MINUS)]`** - (optional) The `-` (minus) character. Only present if the version following it has been yanked since the `versions` file was last recalculated.
+4. **`VERSION_PLATFORM`** - A gem VERSION wich may include the PLATFORM. This combined format is described in more detail in the `info` file format section.
+5. **`[(COMMA)VERSION]`** - (optional) A `,` (comma) character, indicating that another VERSION will follow. Read comma delimited VERSION chunks until a space is encountered.
+6. **`(SPACE)`** - The space character.
+7. **`MD5`** - The MD5 of the gem "info" file, described below. Only the last MD5 for a gem name should be considered accurate for the related info file.
```
Comment (segiddins) [inline rubygems-org-compact-index-api.md:104]:
```suggestion
7. **`MD5`** - The MD5 of the gem "info" file, described below. Only the last MD5 for a gem name in the `versions` file should be considered accurate for the related info file.
```

File: rubygems-org-compact-index-api.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+next: /rubygems-org-rate-limits
+---
+<em class="t-gray">Details of the Compact Index API used for Bundler dependency resolution</em>
+
+Compact Index API
+-----------------
+
+The Compact Index API is considered a stable public API.
+
+The primary index file is the `versions` file, which provides the name, versions and the MD5 checksum of the per-gem `info` files.
```
Comment (segiddins) [inline rubygems-org-compact-index-api.md:15]: there's also the `names` endpoint, which is a simple list of all gem names

**Review by segiddins — APPROVED** [review]

File: rubygems-org-compact-index-api.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+The compact index is designed to be fetched using the HTTP `Range` header.
+When a previously fetched copy is present, a ranged request to take advantage of the appended line pattern.
+
+    Range: bytes=#{range_start}-
+
+Responses from the `/versions` and `/info/[GEM]` endpoints will include the `Repr-Digest` header.
+The digests will have at least the SHA256 checksum of the entire file whether the response includes the full file or a partial response.
+*Please note that a `Digest` header is present, but it is deprecated and may be removed without notice.*
+
+When the Range header is satisfied, append the contents at exactly the starting byte, then computer the SHA256 checksum of the resulting file.
```
Comment (deivid-rodriguez) [inline rubygems-org-compact-index-api.md:46]:
```suggestion
When the Range header is satisfied, append the contents at exactly the starting byte, then compute the SHA256 checksum of the resulting file.
```

File: rubygems-org-compact-index-api.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+The pieces of each line are:
+
+1. **`VERSION`** - The version of the gem. Read VERSION until either `-` (minus) or space character is encountered.
+2. **`[-PLATFORM]`** - The platform, if it is not the default platform `ruby`. The first `-` (minus) character in the `VERSION[-PLATFORM]` chunk splits the VERSION and PLATFORM. The PLATFORM may contain more dashes. Read platform until a space is encountered.
+2. **`(SPACE)`** - The space character.
+3. **`[DEPENDENCY]`** - (optional) A dependency is another gem required by this gem. DEPENDENCY may contain spaces. See below for format.
+5. **`[(COMMA)DEPENDENCY]`** - (optional) A `,` (comma) character, indicating that another DEPENDENCY will follow. Read comma delimited DEPENCENCY chunks until the `|` (pipe) character is encountered.
+6. **`(PIPE)`** - The `|` (pipe) character.
+7. **`METADATA`** - Additional information about the gem, which always includes at least the SHA256 checksum. METADATA may contain spaces. See below for format.
+5. **`[(COMMA)METADATA]`** - (optional) A `,` (comma) character, indicating that another METADATA will follow. Read comma delimited METADATA chunks until the end of the line.
```
Comment (deivid-rodriguez) [inline rubygems-org-compact-index-api.md:148]: Numbering in this list seems odd?

**Review by deivid-rodriguez — APPROVED** [review]
Nice!

File: rubygems-org-compact-index-api.md (hunk 5 above)
Comment (martinemde) [reply rubygems-org-compact-index-api.md:148]: lol, 1 2 2 3 5 6 7 5. Can you guess the next number is this sequence?

File: rubygems-org-compact-index-api.md (hunk 1 above)
Comment (martinemde) [reply rubygems-org-compact-index-api.md:84]: If the dashes are optional and your are reading the file, you can't look for the `---` as the start of the parseable data. Do you just try to read lines until you can parse one? It seems hard to parse reliably if the `---` is optional.

Comment (segiddins) [reply rubygems-org-compact-index-api.md]: https://github.com/rubygems/rubygems/blob/3bcc642d7127cc717c11ebff68d2b459a50e6e4b/bundler/lib/bundler/compact_index_client/cache.rb#L97-L102
